File: NeetCode 150/Arrays/36. Valid Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = [set() for _ in range(9)]
        column = [set() for _ in range(9)]
        grid = [set() for _ in range(9)]

        for r in range(9):
            for c in range (9):
                if(board[r][c] == '.'):
                    continue
                if board[r][c] in row[r]:
                    return False
                else:
                    row[r].add(board[r][c])
                if board[r][c] in column[c]:
                    return False
                else:
                    column[c].add(board[r][c])
                
                grid_index=(r//3)*3 + (c//3)
                if board[r][c] in grid[grid_index]:
                    return False
                else:
                    grid[grid_index].add(board[r][c])
        return True

File: NeedCode 150 (JAVA)/Arrays/36. Valid Sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        HashSet<Character>[] row = new HashSet[9];
        HashSet<Character>[] col = new HashSet[9];
        HashSet<Character>[] grid = new HashSet[9];

        for(int i = 0;i<9;i++){
            row[i]= new HashSet<>();
            col[i] = new HashSet<>();
            grid[i] = new HashSet<>();
        }

        for (int i =0;i<9;i++){
            for(int j =0;j<9;j++){
                if (board[i][j] =='.') continue;
                
                char num = board[i][j];

                if(row[i].contains(num)) return false;
                row[i].add(num);

                if(col[j].contains(num)) return false;
                col[j].add(num);

                int grid_index= (i/3) *3+ (j/3);

                if(grid[grid_index].contains(num)) return false;
                grid[grid_index].add(num);
            }
        }
        return true;

    }
}
